c2corg_api/models/document.py

```diff
@@ -53,15 +67,101 @@ class (e.g. waypoint, route, ...) a separate table will be created, which
+
+    @abc.abstractmethod
+    def to_archive(self):
+        """Create an `Archive*` instance with the same attributes.
+        This method is supposed to be implemented by child classes.
+        """
+        return
+
+    def _to_archive(self, doc):
+        """Copy the attributes of this document into a passed in
+        `Archive*` instance.
+        """
         copy_attributes(self, doc, Document._ATTRIBUTES)
         return doc
 
     def get_archive_locales(self):
         return [locale.to_archive() for locale in self.locales]
 
+    def update(self, other):
+        """Copies the attributes from `other` to this document.
+        Also updates all locales.
+        """
+        copy_attributes(other, self, Document._ATTRIBUTES_WHITELISTED)
+
+        for locale_in in other.locales:
+            locale = self.get_locale(locale_in.culture)
+            if locale:
+                locale.update(locale_in)
+                locale.document_id = self.document_id
+            else:
+                self.locales.append(locale_in)
+
+    def get_versions(self):
+        """Get the version hashs of this document and of all its locales.
+        """
+        version = self.version
+
+        locale_versions = {}
+        for locale in self.locales:
+            locale_versions[locale.culture] = locale.version
+
+        return {
+            'document': version,
+            'locales': locale_versions
+        }
+
+    def get_update_type(self, old_versions):
+        """Get the update type (only figures have changed, only locales have
+        changed, both have changed or nothing) and the languages that have
+        changed.
+        This is done by comparing the old version hashs (before flushing to
+        the database) with the current hashs. Because SQLAlchemy automatically
+        changes the hash, when something has changed, we can easily detect
+        what has changed.
+        """
+        figures_equal = self.version == old_versions['document']
+
+        changed_langs = []
+        locale_versions = old_versions['locales']
+        for locale in self.locales:
+            locale_version = locale_versions.get(locale.culture)
+
+            if not (locale_version and locale_version == locale.version):
+                # new locale or locale has changed
+                changed_langs.append(locale.culture)
+
+        if not figures_equal and changed_langs:
+            return (UpdateType.ALL, changed_langs)
+        elif not figures_equal:
+            return (UpdateType.FIGURES_ONLY, [])
+        elif changed_langs:
+            return (UpdateType.LANG_ONLY, changed_langs)
+        else:
+            return (UpdateType.NONE, [])
+
+    def get_locale(self, culture):
+        """Get the locale with the given culture or `None` if no locale
+        is present.
+        """
+        l = [locale for locale in self.locales if locale.culture == culture]
+        if l:
+            return l[0]
+        else:
+            return None
```

Comment:
So instead of creating a list you are creating an iterator. In this case `self.locales` contains at max 7 entries and the list `l` should contains at max 1 entry. So, memory is not really an argument.

But the iterator version is more concise, I will change it.